File: src/formlabs_local_mcp/preform.py
"""PreFormServer subprocess lifecycle.

When a PreFormServer binary is available locally we start it ourselves, wait
for its `READY FOR INPUT` line, and stop it when the MCP server exits. That
way PreFormServer's unauthenticated HTTP port is only open while an MCP
client is actually connected.
"""

from __future__ import annotations

import asyncio
import contextlib
import logging
import os
import signal
import sys
import time

import httpx

from formlabs_local_mcp.config import Config

log = logging.getLogger(__name__)

READY_TOKEN = "READY FOR INPUT"
DOWNLOAD_URL = "https://formlabs.com/support/Formlabs-API-downloads-and-release-notes"


class PreFormServerProcess:
    def __init__(self, config: Config):
        self._config = config
        self._proc: asyncio.subprocess.Process | None = None
        self._stdout_task: asyncio.Task[None] | None = None
        self._ready = asyncio.Event()

    async def ensure_running(self) -> None:
        """Spawn PreFormServer if configured to, then verify it answers HTTP."""
        cfg = self._config
        if cfg.spawn_preform_server and cfg.preform_server_path:
            if await self._already_reachable():
                log.info("PreFormServer already answering at %s; not spawning", cfg.base_url)
            else:
                await self._spawn()
        await self._wait_until_reachable()

    async def _already_reachable(self) -> bool:
        try:
            async with httpx.AsyncClient(timeout=2.0) as client:
                resp = await client.get(f"{self._config.base_url}/")
                return resp.status_code < 500
        except httpx.HTTPError:
            return False

    async def _spawn(self) -> None:
        cfg = self._config
        path = cfg.preform_server_path
        assert path is not None
        if not path.is_file():
            raise FileNotFoundError(
                f"PreFormServer not found at {path}. Download it from {DOWNLOAD_URL} "
                "and either put PreFormServer.app in /Applications or set "
                "PREFORM_SERVER_PATH to the executable."
            )
        env = dict(os.environ)
        if not cfg.telemetry_enabled:
            env["DISABLE_PREFORMSERVER_TELEMETRY"] = "1"
        log.info("Starting PreFormServer (%s) on port %s", path, cfg.preform_server_port)
        self._proc = await asyncio.create_subprocess_exec(
            str(path),
            "--port",
            str(cfg.preform_server_port),
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.STDOUT,
            stdin=asyncio.subprocess.DEVNULL,
            env=env,
        )
        self._stdout_task = asyncio.create_task(self._drain_stdout())
        try:
            await asyncio.wait_for(self._ready.wait(), timeout=cfg.startup_timeout_seconds)
        except asyncio.TimeoutError as exc:
            await self.shutdown()
            raise RuntimeError(
                f"PreFormServer did not become ready within {cfg.startup_timeout_seconds:.0f}s. "
                "On macOS the first launch can be slow while Gatekeeper verifies the app; "
                "try again, or raise PREFORM_STARTUP_TIMEOUT."
            ) from exc

    async def _drain_stdout(self) -> None:
        assert self._proc is not None and self._proc.stdout is not None
        while True:
            line = await self._proc.stdout.readline()
            if not line:
                return
            text = line.decode(errors="replace").rstrip()
            # stdout is the MCP transport; everything from PreFormServer goes to stderr.
            print(f"[preform] {text}", file=sys.stderr, flush=True)
            if READY_TOKEN in text:
                self._ready.set()

    async def _wait_until_reachable(self) -> None:
        cfg = self._config
        url = f"{cfg.base_url}/"
        deadline = time.monotonic() + 30.0
        async with httpx.AsyncClient(timeout=5.0) as client:
            while True:
                try:
                    resp = await client.get(url)
                    if resp.status_code < 500:
                        return
                except httpx.HTTPError:
                    pass
                if time.monotonic() > deadline:
                    hint = (
                        "Set PREFORM_SERVER_PATH to the PreFormServer executable so it can be "
                        "started automatically, or start it yourself first."
                        if cfg.preform_server_path is None
                        else "PreFormServer was started but never answered HTTP."
                    )
                    raise RuntimeError(f"PreFormServer at {url} is not reachable. {hint}")
                await asyncio.sleep(0.5)

    async def shutdown(self) -> None:
        proc = self._proc
        if proc is None:
            return
        if proc.returncode is None:
            with contextlib.suppress(ProcessLookupError):
                proc.send_signal(signal.SIGTERM)
            try:
                await asyncio.wait_for(proc.wait(), timeout=10.0)
            except asyncio.TimeoutError:
                with contextlib.suppress(ProcessLookupError):
                    proc.kill()
                with contextlib.suppress(asyncio.TimeoutError):
                    await asyncio.wait_for(proc.wait(), timeout=5.0)
        if self._stdout_task:
            self._stdout_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._stdout_task
        self._proc = None

File: src/preform.ts
/**
 * Local PreFormServer lifecycle.
 *
 * We start PreFormServer ourselves on the first tool call, wait for its
 * `READY FOR INPUT` line, and stop it when the MCP server exits. That way its
 * unauthenticated HTTP port is only open while an MCP client is connected.
 */

import { spawn, type ChildProcess } from "node:child_process";
import { existsSync } from "node:fs";
import { createInterface } from "node:readline";
import { type Backend, isReachable, waitUntilReachable } from "./backend.js";
import { DOWNLOAD_PAGE, type Config } from "./config.js";
import { toServerPath } from "./pathmap.js";

export const READY_TOKEN = "READY FOR INPUT";

export function notInstalledMessage(cfg: Config): string {
  return (
    `PreFormServer is not installed. Ask the user for permission, then call the install_preform_server tool ` +
    `(or run: npx -y formlabs-local-mcp install-preform). It downloads PreFormServer from ${DOWNLOAD_PAGE} and verifies ` +
    `Formlabs' code signature. Alternatively set PREFORM_SERVER_PATH to an existing PreFormServer executable. ` +
    `(platform: ${cfg.platform})`
  );
}

export class LocalBackend implements Backend {
  readonly mode = "local" as const;
  private proc: ChildProcess | undefined;

  constructor(private readonly cfg: Config, private readonly log: (line: string) => void = (l) => console.error(l)) {}

  async ensureRunning(): Promise<void> {
    const cfg = this.cfg;
    if (this.proc && this.proc.exitCode === null) {
      if (await isReachable(cfg.baseUrl)) return;
      // Alive but not answering (renderer crash on a headless machine, hung listener): replace it.
      this.log("[preform] PreFormServer is running but not answering; restarting it");
      await this.shutdown();
    }
    if (await isReachable(cfg.baseUrl)) {
      this.log(`[preform] already answering at ${cfg.baseUrl}; not starting another`);
      return;
    }
    if (!cfg.spawn) {
      if (!cfg.preformServerPath && !cfg.env["PREFORM_SERVER_URL"] && cfg.env["PREFORM_SPAWN"] !== "0") throw new Error(notInstalledMessage(cfg));
      await waitUntilReachable(cfg.baseUrl, 3000).catch(() => {
        throw new Error(`PreFormServer at ${cfg.baseUrl} is not reachable. Spawning is disabled (PREFORM_SPAWN=0 or PREFORM_SERVER_URL set), so start PreFormServer yourself and try again.`);
      });
      return;
    }
    const exe = cfg.preformServerPath!;
    if (!existsSync(exe)) throw new Error(`PreFormServer not found at ${exe}. ${notInstalledMessage(cfg)}`);
    await this.spawnServer(exe);
    await waitUntilReachable(cfg.baseUrl, 30_000);
  }

  private async spawnServer(exe: string): Promise<void> {
    const cfg = this.cfg;
    const env: NodeJS.ProcessEnv = { ...process.env };
    if (!cfg.telemetry) env["DISABLE_PREFORMSERVER_TELEMETRY"] = "1";
    if (cfg.launcher.some((w) => /(^|[\\/])wine(64)?$/.test(w))) {
      // Same headless defaults as preform-linux: quiet Wine, no Mono/Gecko prompts, Qt on the bundled software renderer.
      env["WINEDEBUG"] ??= "-all";
      env["WINEDLLOVERRIDES"] ??= "mscoree=d;mshtml=d";
      env["QT_OPENGL"] ??= "software";
    }
    const argv = [...cfg.launcher, exe, "--port", String(cfg.preformServerPort)];
    this.log(`[preform] starting ${argv.join(" ")}`);
    const proc = spawn(argv[0]!, argv.slice(1), { env, stdio: ["ignore", "pipe", "pipe"], windowsHide: true });
    this.proc = proc;
    const ready = new Promise<void>((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new Error(`PreFormServer did not become ready within ${cfg.startupTimeoutMs / 1000}s. On macOS the first launch can be slow while Gatekeeper verifies the app; try again, or raise PREFORM_STARTUP_TIMEOUT.`));
      }, cfg.startupTimeoutMs);
      const onLine = (line: string) => {
        this.log(`[preform] ${line}`); // stdout is the MCP transport; everything from PreFormServer goes to stderr
        if (line.includes(READY_TOKEN)) {
          clearTimeout(timer);
          resolve();
        }
      };
      createInterface({ input: proc.stdout! }).on("line", onLine);
      createInterface({ input: proc.stderr! }).on("line", onLine);
      proc.once("error", (err) => {
        clearTimeout(timer);
        reject(new Error(`Could not start PreFormServer (${argv[0]}): ${err.message}`));
      });
      proc.once("exit", (code) => {
        clearTimeout(timer);
        reject(new Error(`PreFormServer exited with code ${code} before becoming ready`));
      });
    });
    try {
      await ready;
    } catch (err) {
      await this.shutdown();
      throw err;
    }
    proc.once("exit", (code, signal) => {
      if (this.proc === proc) {
        this.log(`[preform] PreFormServer exited unexpectedly (code ${code}, signal ${signal}); it will be restarted on the next call`);
        this.proc = undefined;
      }
    });
  }

  async shutdown(): Promise<void> {
    const proc = this.proc;
    this.proc = undefined;
    if (!proc || proc.exitCode !== null) return;
    const exited = new Promise<void>((r) => proc.once("exit", () => r()));
    proc.kill("SIGTERM");
    const timeout = new Promise<"timeout">((r) => setTimeout(() => r("timeout"), 10_000));
    if ((await Promise.race([exited, timeout])) === "timeout") {
      proc.kill("SIGKILL");
      await Promise.race([exited, new Promise((r) => setTimeout(r, 5000))]);
    }
  }

  stageInput(localPath: string): Promise<string> {
    return Promise.resolve(toServerPath(localPath, this.cfg.pathStyle, this.cfg.pathMap));
  }
  outputPath(localPath: string): Promise<string> {
    return Promise.resolve(toServerPath(localPath, this.cfg.pathStyle, this.cfg.pathMap));
  }
  collectOutput(): Promise<void> {
    return Promise.resolve();
  }
}
